collections/src/main/scala/strawman/collection/mutable/ArrayBuilder.scala

```diff
@@ -0,0 +1,665 @@
+      case _                        => new ArrayBuilder.ofRef[T with AnyRef]()(tag.asInstanceOf[ClassTag[T with AnyRef]]).asInstanceOf[ArrayBuilder[T]]
+    }
+  }
+
+  /** A class for array builders for arrays of reference types.
+   *
+   *  This builder can be reused.
+   *
+   *  @tparam T     type of elements for the array builder, subtype of `AnyRef` with a `ClassTag` context bound.
+   */
+  final class ofRef[T <: AnyRef : ClassTag] extends ArrayBuilder[T] {
+
+    private var elems: Array[T] = _
+    private var capacity: Int = 0
+    private var size: Int = 0
+
+    private def mkArray(size: Int): Array[T] = {
+      val newelems = new Array[T](size)
+      if (this.size > 0) Array.copy(elems, 0, newelems, 0, this.size)
+      newelems
+    }
+
+    private def resize(size: Int): Unit = {
+      elems = mkArray(size)
+      capacity = size
+    }
+
+    override def sizeHint(size: Int): Unit = {
+      if (capacity < size) resize(size)
+    }
+
+    private def ensureSize(size: Int): Unit = {
+      if (capacity < size || capacity == 0) {
+        var newsize = if (capacity == 0) 16 else capacity * 2
+        while (newsize < size) newsize *= 2
+        resize(newsize)
+      }
+    }
+
+    def add(elem: T): this.type = {
+      ensureSize(size + 1)
+      elems(size) = elem
+      size += 1
+      this
+    }
+
+    override def addAll(xs: IterableOnce[T]): this.type = (xs.asInstanceOf[AnyRef]) match {
+      case xs: ArrayView.ofRef[_] =>
+        ensureSize(this.size + xs.length)
+        Array.copy(xs.array, 0, elems, this.size, xs.length)
+        size += xs.length
+        this
+      case _ =>
+        super.addAll(xs)
+    }
+
+    def clear(): Unit = size = 0
+
+    def result() = {
+      if (capacity != 0 && capacity == size) {
+        capacity = 0
+        elems
+      }
+      else mkArray(size)
+    }
+
+    override def equals(other: Any): Boolean = other match {
+      case x: ofRef[_] => (size == x.size) && (elems == x.elems)
+      case _ => false
+    }
+
+    override def toString = "ArrayBuilder.ofRef"
+  }
+
+  /** A class for array builders for arrays of `byte`s. It can be reused. */
+  final class ofByte extends ArrayBuilder[Byte] {
+
+    private var elems: Array[Byte] = _
+    private var capacity: Int = 0
+    private var size: Int = 0
+
+    private def mkArray(size: Int): Array[Byte] = {
+      val newelems = new Array[Byte](size)
+      if (this.size > 0) Array.copy(elems, 0, newelems, 0, this.size)
+      newelems
+    }
+
+    private def resize(size: Int): Unit = {
+      elems = mkArray(size)
+      capacity = size
+    }
+
+    override def sizeHint(size: Int): Unit = {
+      if (capacity < size) resize(size)
+    }
+
+    private def ensureSize(size: Int): Unit = {
+      if (capacity < size || capacity == 0) {
+        var newsize = if (capacity == 0) 16 else capacity * 2
+        while (newsize < size) newsize *= 2
+        resize(newsize)
+      }
+    }
```

Comment:
Do you have an idea of the loss of performance we would get if we defined these methods in `ArrayBuilder`?

Reply:
I haven't benchmarked it but I wouldn't expect any performance difference. `ensureSize` and `sizeHint` don't touch the actual array. They should be OK to pull up.